src/raftstore/store/snap/builder.rs

```diff
@@ -0,0 +1,558 @@
+
+            let cf_builder = inner.tmp_cf_files.last_mut().unwrap();
+            snap_key_count += cf_builder.scan_cf(
+                &db_snap,
+                region.get_id(),
+                &start_key,
+                &end_key,
+                inner.io_limiter.as_ref().map(|arc| arc.as_ref()),
+                inner.key,
+                self.snap_stale_notifier.as_ref(),
+            )?;
+        }
+
+        // Rename all tmp files to cf files.
+        let snapshot_meta = inner.save()?;
+        let total_size = get_size_from_snapshot_meta(&snapshot_meta);
+
+        SNAPSHOT_BUILD_TIME_HISTOGRAM.observe(duration_to_sec(t.elapsed()) as f64);
+        SNAPSHOT_KV_COUNT_HISTOGRAM.observe(snap_key_count as f64);
+        SNAPSHOT_SIZE_HISTOGRAM.observe(total_size as f64);
+        info!(
+            "[region {}] scan snapshot {}, size {}, key count {}, takes {:?}",
+            region.get_id(),
+            gen_display_path(&inner.dir, inner.for_send, inner.key),
+            total_size,
+            snap_key_count,
+            t.elapsed(),
+        );
+
+        self.size_tracker.fetch_add(total_size, Ordering::SeqCst);
+        Ok(snapshot_meta)
+    }
+}
+
+pub struct SnapshotReceiver {
+    inner: SnapshotBase,
+    snapshot_meta: SnapshotMeta,
+    size_tracker: Arc<AtomicU64>,
+    cur_cf_pos: usize,
+}
+
+impl SnapshotReceiver {
+    pub(super) fn new(
+        dir: String,
+        key: SnapKey,
+        io_limiter: Option<Arc<IOLimiter>>,
+        snapshot_meta: SnapshotMeta,
+        size_tracker: Arc<AtomicU64>,
+    ) -> SnapshotReceiver {
+        let inner = SnapshotBase::new(dir, false, key, io_limiter);
+        SnapshotReceiver {
+            inner,
+            snapshot_meta,
+            size_tracker,
+            cur_cf_pos: 0,
+        }
+    }
+
+    pub fn save(&mut self) -> Result<()> {
+        self.flush()?;
+        let mut got_meta = self.inner.save()?;
+        let got_total_size = get_size_from_snapshot_meta(&got_meta);
+        let total_size = get_size_from_snapshot_meta(&self.snapshot_meta);
+        if got_total_size != total_size {
+            self.inner.success = false;
+            return Err(snapshot_size_corrupt(total_size, got_total_size));
+        }
+
+        let cf_files = got_meta.take_cf_files().into_iter();
+        for (got_cf, want_cf) in cf_files.zip(self.snapshot_meta.get_cf_files()) {
+            let expected = want_cf.get_size();
+            let got = got_cf.get_size();
+            if expected != got {
+                self.inner.success = false;
+                return Err(snapshot_size_corrupt(expected, got));
+            }
+
+            let expected = want_cf.get_checksum();
+            let got = got_cf.get_checksum();
+            if expected != got {
+                self.inner.success = false;
+                return Err(snapshot_checksum_corrupt(expected, got));
+            }
+        }
+
+        self.size_tracker.fetch_add(total_size, Ordering::SeqCst);
```

Comment:
where to set `inner.success = true`?

Reply:
`innser.success` is set `true` when `Inner::save` success.